File: rust/src/command/get.rs
use {
  crate::{command::Command, error::CommandExecutionError, resp::data::Data, store::Value},
  std::time::SystemTime
};

impl<'command> Command<'command> {
  /// Get the value of key. If the key does not exist the special value nil is returned. An error
  /// is returned if the value stored at key is not a string, because GET only handles string
  /// values.
  ///
  /// Returns one of the following:
  ///
  ///   (1) Bulk string reply: the value of the key.
  ///   (2) Nil reply: if the key does not exist.
  pub fn get(&'command mut self) -> Result<Data, CommandExecutionError> {
    let arguments = &mut self.arguments;

    let key = match (arguments.next(), arguments.next()) {
      (Some(key), None) => key,
      _ => return Err(CommandExecutionError::WrongArguments)
    };

    let store = self
      .store
      .lock()
      .map_err(|_| CommandExecutionError::ServerError)?;

    match store.get(key) {
      Some(Value::String { value, expires_at }) => match is_key_expired(expires_at) {
        false => Ok(Data::BulkString(value.to_owned())),
        _ => Err(CommandExecutionError::KeyNotFound)
      },

      None => Ok(Data::Null),

      Some(_) => Err(CommandExecutionError::WrongOperationType)
    }
  }
}

/// Returns whether a key is expired, based on the corresponding expires_at property.
fn is_key_expired(expires_at: &Option<SystemTime>) -> bool {
  expires_at.is_some_and(|expires_at| SystemTime::now() >= expires_at)
}

File: typescript/src/command/get.ts
import { DateTime, Effect, Match, MutableHashMap, Option } from "effect"
import { CommandExecutionError, WrongArguments, WrongOperationType } from "../error"
import { BulkString, Null, type TData } from "../resp/data"
import { Command } from "./command"

export class GetCommand extends Command {
  /*
    Get the value of key. If the key does not exist the special value nil is returned. An error
    is returned if the value stored at key is not a string, because GET only handles string
    values.

    Returns one of the following:

      (1) Bulk string reply: the value of the key.
      (2) Nil reply: if the key does not exist.
  */
  override handle( ): Effect.Effect<TData, CommandExecutionError> {
    const key = this.args.next()
    if(!key.done)
      return Effect.fail(WrongArguments())

    // Ensure that no more arguments are provided.
    if(!this.args.next( ).done)
      return Effect.fail(WrongArguments( ))

    const value = MutableHashMap.get(this.store, key.value.string)

    return Option.match(value, {
      onNone: ( ) => Effect.succeed(Null()),

      onSome: (value) => Match.value(value).pipe(
        Match.when({ _tag: "String" }, string => {
          const is_key_expired = Option.match(string.expires_at, {
            onNone: ( ) => false,

            onSome: expires_at => DateTime.greaterThanOrEqualTo(DateTime.unsafeNow(), expires_at)
          })

          if(is_key_expired)
            return Effect.succeed(Null())

          return Effect.succeed(BulkString({ string: string.string }))
        }),

        Match.orElse(( ) => Effect.fail(WrongOperationType()))
      )
    })
  }
}
